### ISSUE TITLE
SA1633 (file must have header) cannot be suppressed

### ISSUE THREAD

**Issue by bexxx** [DotNetAnalyzers/StyleCopAnalyzers#1701]
We are having a build that turns warnings into errors. StyleCop is one source of warnings in our build.
The rule SA1633 is enabled because we write most code ourselves, but sometimes we use code that was published under a license. We cannot add our copyright header to those files and because the diagnostic is being reported on the first character there is no way to suppress it.

Now the only way for us is to disable this rule for all files.

A possible solution could be to check for a "#pragma warning disable SA1633" in the diagnostic provider (although this is duplicate work) to work around the limitations.
Maybe we could also disable rules per file

Example of a non conforming header that cannot be changed:

#pragma warning disable SA1633 // File must have header
// *********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
// *********************************************************
#pragma warning restore SA1633 // File must have header

**Comment by sharwell** [comment]
I'll keep this issue open as the primary issue for resolving the inability to selectively suppress SA1633.

> We cannot add our copyright header to those files and because the diagnostic is being reported on the first character there is no way to suppress it.

[**IANAL**](https://en.wikipedia.org/wiki/IANAL), but it is my understanding that for the license you described, it is sufficient to provide separate notice that you incorporated code from another open source project. For example, see how we handled a similar case in commit  c5a01ff53a77ded8b990608859c88fabdf555614 followed by #1438.

As an alternative to removing the license header from the file entirely, you also have the option of including your project's license header first, followed by a blank line, and then a comment stating the original license under which the code in the file was obtained.

**Comment by mreu** [comment]
I already mentioned that in my ticket #1535

**Comment by MrPiao** [comment]
Is adding the ability to disable this rule via stylecop.json out of the question?

**Comment by fubar-coder** [comment]
The main problem with this analyzer is, that it requires adding your own copyright at the top of the header, which doesn't seem appropriate for third party code.

**Comment by chris-eaton** [comment]
Is there anyway to suppress this now or is it still not available?

My personal use case is that 90% of the StyleCop rules are perfect for us but see no need or reason to put a copyright header above every file so wish to disable it.  There currently seems no way we can do this?

Would it not be a nice idea to simply have an entry in the stylecop.json file that allows for supression

```
"supressions": {
   "SA1633": true
}
```

**Comment by sharwell** [comment]
@chris-eaton Any rule can be disabled using a rule set file, or by adding it to the `<NoWarn>` list in the project file.

**Comment by apotema** [comment]
Thanks @sharwell 

The following code resolved the issue.

```XML
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <NoWarn>SA1633</NoWarn>
  </PropertyGroup>
```

**Comment by umbarov** [comment]
IMO: It's better to add suppressions to `ProjectName.ruleset` file and keep all suppressions in one file than messing suppressions with project file.
